### PR TITLE
HDDS-8997. Fix RangerOzoneAuthorizer initialization issue.

### PR DESCRIPTION
**Pull request by ashishkumar50**
## What changes were proposed in this pull request?

External access authorizer can be shared in same service. In case of ranger we can see we initialize multiple times in same service which eventually does refresh policies internally inside ranger plugin. Any request at this point doesn't work as policy check returns null. Which can be avoided by initialize only once during OM startup. 

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8997

## How was this patch tested?

Run existing tests.

### CODE REVIEW

**Comment by ashishkumar50** [comment]
@smengcl, Please help to review.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataReader.java [hunk 1]
```diff
@@ -104,19 +105,29 @@ public OmMetadataReader(KeyManager keyManager,
...
-        authorizer.setKeyManager(keyManager);
-        authorizer.setPrefixManager(prefixManager);
-        authorizer.setAdminCheck(ozoneManager::isAdmin);
-        authorizer.setReadOnlyAdminCheck(ozoneManager::isReadOnlyAdmin);
-        authorizer.setAllowListAllVolumes(allowListAllVolumes);
+      Class<? extends IAccessAuthorizer> clazz = configuration.getClass(
+          OZONE_ACL_AUTHORIZER_CLASS, OzoneAccessAuthorizer.class,
+          IAccessAuthorizer.class);
+      if (bNeedAccessAuthInitialization ||
+          clazz.getSimpleName().equals("OzoneNativeAuthorizer")) {
```
Comment (smengcl) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataReader.java:112]:
```suggestion
          clazz.getSimpleName().equals(OzoneNativeAuthorizer.class.getSimpleName())) {
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataReader.java [hunk 2]
```diff
@@ -104,19 +105,29 @@ public OmMetadataReader(KeyManager keyManager,
...
+          authorizer.setKeyManager(keyManager);
+          authorizer.setPrefixManager(prefixManager);
+          authorizer.setAdminCheck(ozoneManager::isAdmin);
+          authorizer.setReadOnlyAdminCheck(ozoneManager::isReadOnlyAdmin);
+          authorizer.setAllowListAllVolumes(allowListAllVolumes);
+        } else {
+          isNativeAuthorizerEnabled = false;
+        }
       } else {
+        accessAuthorizer = ozoneManager.getAccessAuthorizer();
```
Comment (smengcl) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataReader.java:130]: Looks like `accessAuthorizer` isn't closed or destroyed upon `OmMetadataReader` instance closure. Should be fine to share it like this IMO.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataReader.java (hunk 1 above)
Comment (ashishkumar50) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataReader.java]: Done

**Review by smengcl — APPROVED** [review]
lgtm pending CI

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataReader.java [hunk 3]
```diff
@@ -90,7 +90,8 @@ public OmMetadataReader(KeyManager keyManager,
                           OzoneManager ozoneManager,
                           Logger log,
                           AuditLogger audit,
-                          OmMetadataReaderMetrics omMetadataReaderMetrics) {
+                          OmMetadataReaderMetrics omMetadataReaderMetrics,
+                          boolean bNeedAccessAuthInitialization) {
```
Comment (adoroszlai) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataReader.java:94]: Nit: please don't prefix variables with type.

**Review by adoroszlai — CHANGES_REQUESTED** [review]
Thanks @ashishkumar50 for working on this.  While the patch is fine to minimize change, I think it would be better to let `OzoneManager` create and store `accessAuthorizer` just before creation of `OmMetadataReader`:

https://github.com/apache/ozone/blob/300e7a76481480ae2651ff7214d7b9f146091e68/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java#L828-L829

Then `OmMetadataReader` could simply store reference to `ozoneManager.getAccessAuthorizer()` without complex logic.

https://github.com/apache/ozone/blob/300e7a76481480ae2651ff7214d7b9f146091e68/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataReader.java#L96-L97

**Comment by ashishkumar50** [comment]
> Thanks @ashishkumar50 for working on this. While the patch is fine to minimize change, I think it would be better to let `OzoneManager` create and store `accessAuthorizer` just before creation of `OmMetadataReader`:
> 
> https://github.com/apache/ozone/blob/300e7a76481480ae2651ff7214d7b9f146091e68/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java#L828-L829
> 
> Then `OmMetadataReader` could simply store reference to `ozoneManager.getAccessAuthorizer()` without complex logic.
> 
> https://github.com/apache/ozone/blob/300e7a76481480ae2651ff7214d7b9f146091e68/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataReader.java#L96-L97

@adoroszlai, Thanks for the review, We can create and store `accessAuthorizer` in `OzoneManager`, but problem is for `nativeAuthorizer` which always need new instance.

**Comment by adoroszlai** [comment]
> problem is for `nativeAuthorizer` which always need new instance.

Ah, OK, sorry, I missed that part.

**Comment by adoroszlai** [comment]
Thanks @ashishkumar50 for the patch, @smengcl for the review.

**Pull request merged**